File: src/main/java/ley/anvil/modpacktools/commands/DownloadMods.java
package ley.anvil.modpacktools.commands;

import ley.anvil.modpacktools.Main;
import ley.anvil.modpacktools.command.CommandReturn;
import ley.anvil.modpacktools.command.ICommand;
import ley.anvil.modpacktools.command.LoadCommand;
import ley.anvil.modpacktools.util.FileDownloader;
import ley.anvil.modpacktools.util.Util;

import java.io.File;
import java.net.MalformedURLException;
import java.net.URL;
import java.nio.file.Paths;
import java.util.Arrays;
import java.util.function.Function;
import java.util.stream.Collectors;

//TODO due to tilera still not having a proper implementation to get mod info in the addonscript library, this always downloads the most recent version of mods!
@LoadCommand
public class DownloadMods implements ICommand {
    @Override
    public CommandReturn execute(String[] args) {
        if(args.length < 2)
            return CommandReturn.fail("Invalid args");

        Function<String, URL> toURL = spec -> {
            try {
                return new URL(spec);
            }catch(MalformedURLException ignored) {
            }
            return null;
        };

        FileDownloader.downloadAsync(
                Main.MPJH.getJson().getDefaultVersion().getRelLinks(Main.MPJH.getJson().indexes, "client", false, "internal.dir:mods", null).stream()
                        .collect(Collectors.toMap(
                                i -> Util.sanitizeURL(toURL.apply(i.getKey())),
                                i -> new File(args[1], Paths.get(toURL.apply(i.getKey()).getPath()).getFileName().toString())
                        )),
                r -> {
                    //Synced to prevent the exception being printed too late
                    synchronized(this) {
                        System.out.println(r.getResponseCode() + " " + r.getResponseMessage() + " " + r.getUrl() + " " + r.getFile());
                        if(r.getException() != null)
                            System.out.println(r.getException().getMessage());
                    }
                },
                Arrays.asList(args).contains("force") ? FileDownloader.AsyncDownloader.ExistingFileBehaviour.OVERWRITE : FileDownloader.AsyncDownloader.ExistingFileBehaviour.SKIP
        );
        return CommandReturn.success();
    }

    @Override
    public String getName() {
        return "downloadmods";
    }

    @Override
    public boolean needsConfig() {
        return true;
    }

    @Override
    public String getHelpMessage() {
        return "Downloads all mods. force always downloads files even if they are already present Syntax: <OutDir> [force]";
    }
}

File: src/main/kotlin/ley/anvil/modpacktools/commands/DownloadMods.kt
package ley.anvil.modpacktools.commands

import ley.anvil.addonscript.wrapper.FileOrLink
import ley.anvil.modpacktools.MPJH
import ley.anvil.modpacktools.command.AbstractCommand
import ley.anvil.modpacktools.command.CommandReturn
import ley.anvil.modpacktools.command.CommandReturn.Companion.success
import ley.anvil.modpacktools.command.LoadCommand
import ley.anvil.modpacktools.util.FileToDownload
import ley.anvil.modpacktools.util.arg
import ley.anvil.modpacktools.util.downloadFiles
import ley.anvil.modpacktools.util.fPrintln
import net.sourceforge.argparse4j.impl.Arguments.storeTrue
import net.sourceforge.argparse4j.impl.type.FileArgumentType
import net.sourceforge.argparse4j.inf.ArgumentParser
import net.sourceforge.argparse4j.inf.Namespace
import java.io.File

@LoadCommand
object DownloadMods : AbstractCommand("DownloadMods") {
    override val helpMessage: String = "Downloads all mods."

    override fun ArgumentParser.addArgs() {
        arg("dir") {
            type(FileArgumentType().verifyCanCreate())
            help("the directory to download the mods to")
        }

        arg("-f", "--force") {
            action(storeTrue())
            help("if true, mods that are already in the download folder will be downloaded again")
        }

        arg("-a", "--all") {
            action(storeTrue())
            help("Downloads not only mods but everything with a dir installer")
        }
    }

    override fun execute(args: Namespace): CommandReturn {
        val json = MPJH.asWrapper
        val fileList = mutableListOf<FileOrLink>()
        for(rel in json!!.defaultVersion!!.getRelations {"client" in it.options && (args.getBoolean("all") || it.type == "mod")})
            if(rel.hasFile())
                fileList.add(rel.file.get())

        downloadFiles(
            fileList
                .filter {it.isURL}
                .filter {
                    val (installer, dir) = it.installer.split(':')

                    installer == "internal.dir" && (args.getBoolean("all") || dir == "mods")
                }
                .map {
                    FileToDownload(
                        if(args.getBoolean("all"))
                            File(args.get<File>("dir"), it.installer.split(':')[1])
                        else
                            args.get("dir"),
                        it.url,
                        true,
                        !args.getBoolean("force")
                    )
                }
        ) {
            println("${it.responseCode} ${it.responseMessage} ${it.file.url} ${it.downloadedFile}")
            if(it.exception != null) {
                fPrintln("ERROR DOWNLOADING ${it.file.url}")
                it.exception.printStackTrace()
            }
        }
        return success()
    }
}
